.github/workflows/main.yml

```diff
@@ -126,66 +114,137 @@ jobs:
+        include:
+          - image: macos-10.15
+            platform: macos
+          - image: windows-2019
+            platform: windows
+          - image: ubuntu-16.04  # Old Ubuntu version for old glibc
+            platform: linux
+    runs-on: ${{ matrix.image }}
+    env:
+      CI_BUILD_WHEEL: ${{ matrix.platform == 'linux' }}
+      CI_BUILD_PYINSTALLER: 1
+      CI_BUILD_WININSTALLER: ${{ matrix.platform == 'windows' }}
+      CI_BUILD_KEY: ${{ secrets.CI_BUILD_KEY }}
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          persist-credentials: false
+          fetch-depth: 0
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.9'
+      - if: matrix.platform == 'windows'
+        uses: actions/cache@v2
+        with:
+          path: release/installbuilder/setup
+          key: installbuilder
+      - run: python release/pins/pip-install-pinned.py
+      - run: python release/cibuild.py build
+      # artifacts must have different names, see https://github.com/actions/upload-artifact/issues/24
+      - uses: actions/upload-artifact@v2
+        with:
+          name: binaries.${{ matrix.platform }}
+          path: release/dist
 
   # Separate from everything else because slow.
   build-and-deploy-docker:
     if: github.repository == 'mitmproxy/mitmproxy' && github.event_name == 'push'
     environment: deploy-docker
-    needs: [test, test-web-ui, build-wheel]
+    needs: [ test, test-web-ui, build ]
     runs-on: ubuntu-latest
     env:
       CI_BUILD_DOCKER: 1
       DOCKER_USERNAME: mitmbot
       DOCKER_PASSWORD: ${{ secrets.DOCKER_PASSWORD }}
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/checkout@v2
+        with:
+          persist-credentials: false
       - uses: actions/setup-python@v2
         with:
           python-version: '3.9'
-      - run: pip install tox
       - uses: actions/download-artifact@v2
         with:
-          name: wheel
+          name: build.linux
           path: release/dist
-      - run: tox -e cibuild -- build
-      - run: tox -e cibuild -- upload
+      - run: python release/pins/pip-install-pinned.py
+      - run: python release/cibuild.py build
+      - run: python release/cibuild.py upload
 
   deploy:
     if: github.repository == 'mitmproxy/mitmproxy' && github.event_name == 'push'
     environment: deploy
     runs-on: ubuntu-latest
-    needs: [test, test-web-ui, build-wheel, build-binaries]
+    needs: [test, test-web-ui, build]
     env:
       CI_BUILD_WHEEL: 1
       CI_BUILD_PYINSTALLER: 1
       CI_BUILD_WININSTALLER: 1
       TWINE_USERNAME: mitmproxy
-      TWINE_PASSWORD: ${{ secrets.TWINE_PASSWORD }}
-      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
-      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+      #TWINE_PASSWORD: ${{ secrets.TWINE_PASSWORD }}
+      #AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+      #AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
```

Comment:
TODO: Figure out if this needs to be explicitly specified if `environment` is set.